resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/KubernetesClientBuilder.scala

```diff
@@ -32,12 +34,13 @@ private[spark] object KubernetesClientBuilder {
    * are picked up from canonical locations, as they are injected
    * into the pod's disk space.
    */
-  def buildFromWithinPod(
-      kubernetesMaster: String,
-      kubernetesNamespace: String): DefaultKubernetesClient = {
+  def buildFromWithinPod(kubernetesNamespace: String, useSsl: Boolean): DefaultKubernetesClient = {
+    val kubernetesHost = System.getenv(ENV_KUBERNETES_SERVICE_HOST)
+    val kubernetesPort = System.getenv(ENV_KUBERNETES_SERVICE_PORT)
+    val urlScheme = if (useSsl) "https" else "http"
     var clientConfigBuilder = new ConfigBuilder()
       .withApiVersion("v1")
-      .withMasterUrl(kubernetesMaster)
+      .withMasterUrl(s"$urlScheme://$kubernetesHost:$kubernetesPort")
```

Comment:
This should be `https` always. Even if the user uses the insecure endpoint to access the apiserver from outside the cluster, KUBENRETES_SERVICE_PORT should point to the secure endpoint.